tests/test_ingest.py

```diff
@@ -0,0 +1,154 @@
+"""Unit tests for ingest.py
+
+These tests verify that various matrix file types can be extracted and
+transformed, as expected by code that loads transformed data into Firestore.
+
+Mocks are used for test speed and isolation.
+
+PREREQUISITE
+Spin up Python 3.6 virtualenv, install Python dependencies in requirements.txt
+Also, before "EXAMPLES" commands, run `cd tests`.
+
+EXAMPLES
+
+# Run all tests
+pytest
+
+# Run tests with names containing the string "dense"
+pytest -k "dense"
+
+# Run all tests in a manner that shows any print() statements
+python3 test_ingest.py
+
+# Run all tests, using multiple CPUs
+# Details: https://pypi.org/project/pytest-xdist/
+pytest -n auto
+
+# Run all tests, show code coverage metrics
+# For explanation of coverage report, see:
+# https://coverage.readthedocs.io/en/v4.5.x/branch.html
+coverage run --branch test_ingest.py; coverage report -m --include *scp-ingest-pipeline/ingest*
+
+"""
+
+import ast
+from glob import glob
+import sys
+import unittest
+from unittest.mock import patch
+
+import google.cloud
+
+sys.path.append('../ingest')
+from ingest_pipeline import *
+
+def mock_load_expression_data(self, *args, **kwargs):
+    """Enables overwriting normal function with this placeholder.
+    Returning the arguments enables tests to verify that the code invokes
+    this method with expected argument values.
+
+    Ideally we would mock Firestore itself, but existing libraries to mock
+    Firestore (e.g. https://github.com/mdowds/python-mock-firestore) don't
+    implement important Firestore methods like "batch()".
+    """
+    self.load_expression_data_args = args
+    self.load_expression_data_kwargs = kwargs
+
+def mock_firestore_client(*args, **kwargs):
```

Comment:
I don't think this function is required. You're not expecting a returned value, if anything you're testing to see that it was called with the correct arguments.  I think you could do something like firestore.Client = Mock() and mock the whole object. That way when you call any method of firestore.Client it would just returned a mocked object.

Reply:
> if anything you're testing to see that it was called with the correct arguments

That's not the intent.  

This suite does not test that `firestore.Client()` itself gets expected arguments (it wouldn't; the [code doesn't call it with arguments](https://github.com/broadinstitute/scp-ingest-pipeline/blob/63f0ba61e3f30178b9396a2a5c41592d8e3f7193/ingest/ingest_pipeline.py#L62)), it tests that `IngestPipeline.load_expression_data` gets expected arguments.  The tests make assertions about the code slightly upstream of where (and more granularly than) they ideally would.  The reason, as explained in the docstring starting on line 46:

"Ideally we would mock Firestore itself, but existing libraries to mock Firestore (e.g. https://github.com/mdowds/python-mock-firestore) don't implement important Firestore methods like 'batch()'."

> I think you could do something like firestore.Client = Mock() and mock the whole object

`mock_firestore_client` does mock the whole object, albeit superficially.  If we added `firestore.Client = Mock()`, we'd add it to this patch method as recommended [here](https://stackoverflow.com/questions/21982964/how-to-get-the-call-count-using-mock-patch).  However, while that would enable slightly greater coverage, it would also be more much brittle as it would entail testing more even granular implementation details than done now.  The benefit of greater coverage would be exceeded by the cost of lower test maintainability.

Ideally, we would test not just the high-level input to our loading code as done now, but also the high-level output.

For robust value beyond such very thin mocks, we would need to implement `batch()` and any other missing methods in [`python-mock-firestore`](https://github.com/mdowds/python-mock-firestore).  Then we could make assertions about the state of a free, fast, and realistic in-memory Firestore mock in a way that is high-coverage and durable across internal code changes.

Reply:
We should mock all of firestore or firestore.Client when we utilize their functions regardless of what suite we're using. When you mock the client object any function you call from it returns a mocked object. This is because when you mock firestore.Client, all functions of the Client object become mocked because it's implied.